[lldb][progress] Correctly check total for deterministic progress

chelcassanova opened the pull request:
The `total` parameter for the constructor for Progress was changed to a std::optional in https://github.com/llvm/llvm-project/pull/77547. When initializing the `m_total` member variable for progress, it is set to 1 if the `total` parameter is std::nullopt. Other areas of the code were still checking if `m_total` was a UINT64_MAX to determine if the progress was deterministic or not, so these have been changed to check for the integer 1.

The member variable `m_total` could be changed to a std::optional as well, but this means that the `ProgressEventData::GetTotal()` (which is used for the public API) would
either need to return a std::optional value or it would return some specific integer to represent non-deterministic progress if `m_total` is std::nullopt.

chelcassanova committed:
[lldb][progress] Correctly check total for deterministic progress

The `total` parameter for the constructor for Progress was changed to a
std::optional in https://github.com/llvm/llvm-project/pull/77547. When
initializing the `m_total` member variable for progress, it is set to 1
if the `total` parameter is std::nullopt. Other areas of the code were
still checking if `m_total` was a UINT64_MAX to determine if the
progress was deterministic or not, so these have been changed to check
for the integer 1.

The member variable `m_total` could be changed to a std::optional as
well, but this means that the `ProgressEventData::GetTotal()` (which is
used for the public API) would
either need to return a std::optional value or it would return some
specific integer to represent non-deterministic progress if `m_total`
is std::nullopt.

medismailben reviewed (approved):
LGTM!

clayborg commented on lldb/include/lldb/Core/DebuggerEvents.h:
```diff
@@ -39,7 +39,7 @@ class ProgressEventData : public EventData {
 
   uint64_t GetID() const { return m_id; }
-  bool IsFinite() const { return m_total != UINT64_MAX; }
+  bool IsFinite() const { return m_total != 1; }
```
It would be nice to actually hard code a constant value inside the progress class and use it. We are using magic numbers.
```
class Progress {
constexpr uint64_t kNonDeterministicTotal = UINT64_MAX;
```
I was also thinking that UINT64_MAX might be a better value to use than 1 because people can create valid deterministic Progress with 1 work item. So this code would be:
```
bool IsFinite() const { return m_total != kNonDeterministicTotal; }
```

clayborg replied:
I realize I had suggested using the magic number 1 in previous reviews...

chelcassanova replied:
I added your suggestions to the PR, no longer using a magic number or the number 1 to represent non-deterministic progress

clayborg commented on lldb/source/Core/DebuggerEvents.cpp:
```diff
@@ -41,7 +41,7 @@ void ProgressEventData::Dump(Stream *s) const {
   // If m_total is UINT64_MAX, there is no progress to report, just "start"
   // and "end". If it isn't we will show the completed and total amounts.
-  if (m_total != UINT64_MAX)
+  if (m_total != 1)
```
```
if (m_total != Progress::kNonDeterministicTotal)
```

clayborg commented on lldb/source/Core/Progress.cpp:
```diff
@@ -23,7 +23,6 @@ Progress::Progress(std::string title, std::string details,
                    lldb_private::Debugger *debugger)
     : m_title(title), m_details(details), m_id(++g_id), m_completed(0),
       m_total(1) {
```
```
m_total(kNonDeterministicTotal) {
```

JDevlieghere reviewed (approved):
LGTM with Greg's suggestion.

chelcassanova committed:
Use UINT64_MAX, use static class var instead of magic number

Uses a static class member set to UINT64_MAX instead of using a magic
number of 1 to represent non-deterministic progress.

clayborg reviewed (approved):
Looks good!
